## Overview

This analysis is part of a broader project focused on understanding Ethereum gas fee dynamics using a lightweight, sampled data collection pipeline. Because full-chain indexing is computationally expensive, the dataset for this study is intentionally based on sampling rather than complete blockchain coverage.

Each day, the pipeline collected:

* 50 randomly selected blocks

* Up to 20 sampled transactions per block

This results in approximately 1,000 transactions and 50 blocks per day, representing a small but consistent snapshot of Ethereum’s on-chain activity. The aim of this sampling strategy is not to recreate the full network state, but to:

* Capture relative daily trends

* Analyze gas price behavior

* Observe whether congestion signals appear

* Understand user cost dynamics (in gwei and USD)

* Compare network demand indicators (utilization, tx count, value moved)

Given the small sample size, the results are expected to show higher noise, especially in metrics like total value transferred or transaction count. However, sampled data is still sufficient to identify broad behavioral patterns, such as:

* Whether gas prices rise with utilization

* Whether activity affects congestion

* Whether USD fees track ETH price

* Whether weekends differ from weekdays

* Whether any sustained congestion occurs

The analysis should therefore be interpreted as a trend-oriented study, not a full-network reconstruction. Its primary goal is to evaluate how Ethereum’s gas fee mechanisms behave under typical low-demand periods using a sampled dataset, aligned with the design of the larger ETL and analytics pipeline described in the project specification.

In [ ]:
import pandas as pd
from sqlalchemy import create_engine

# this method is recommended when i am using pandas and jupyter for connecting to the database
engine = create_engine("postgresql://user:password@localhost:port/project_name")

df = pd.read_sql("SELECT * FROM daily_network_stats ORDER BY day", engine)
df.head()

day  sampled_block_count  daily_block_tx_count  sampled_tx_count  \
0  2025-11-01                   50                  9755              1000   
1  2025-11-02                   50                 12156              1000   
2  2025-11-03                   50                 10886              1000   
3  2025-11-04                   50                 11937              1000   
4  2025-11-05                   50                 10586              1000   

   median_gas_fee_gwei  median_gas_fee_eth  median_gas_fee_usd  \
0         12803.016247            0.000013            0.049598   
1         13174.317219            0.000013            0.051150   
2         52502.874954            0.000053            0.194858   
3         94952.750472            0.000095            0.329891   
4         51124.413695            0.000051            0.172761   

   avg_tx_fee_eth  avg_tx_fee_usd  daily_block_gas_fee_eth  \
0        0.000069        0.267974                 0.069261   
1        0.000067        0.258604                 0.066572   
2        0.000222        0.810683                 0.221769   
3        0.000498        1.667584                 0.498130   
4        0.000205        0.679235                 0.204745   

   daily_block_gas_fee_usd  avg_block_gas_price_gwei  avg_block_utilization  \
0               267.974114                  0.898853               0.578981   
1               258.604047                  0.811082               0.536854   
2               810.682864                  2.332519               0.504186   
3              1667.583628                  3.769551               0.527059   
4               679.235086                  1.482445               0.487917   

   avg_eth_price_usd  max_eth_price_usd  min_eth_price_usd  total_value_eth  \
0        3870.442917            3895.15            3844.17     1.999548e+22   
1        3875.721250            3908.24            3852.86     3.972325e+20   
2        3693.525000            3865.11            3587.97     5.854476e+20   
3        3460.027083            3644.58            3213.52     7.623283e+20   
4        3360.889167            3462.04            3240.47     1.142191e+22   

   total_value_usd  
0     7.732523e+25  
1     1.537361e+24  
2     2.148277e+24  
3     2.485302e+24  
4     3.794438e+25

In [5]:
df.describe()

sampled_block_count  sampled_tx_count  median_gas_fee_gwei  \
count            31.000000         31.000000            31.000000   
mean             49.967742        998.129032         30444.455939   
std               0.179605          5.371170         22653.055280   
min              49.000000        980.000000          5018.757470   
25%              50.000000       1000.000000         12717.119376   
50%              50.000000       1000.000000         23116.531872   
75%              50.000000       1000.000000         42219.515518   
max              50.000000       1000.000000         94952.750472   

       median_gas_fee_eth  median_gas_fee_usd  avg_tx_fee_eth  avg_tx_fee_usd  \
count           31.000000           31.000000       31.000000       31.000000   
mean             0.000030            0.099156        0.000149        0.473116   
std              0.000023            0.076620        0.000106        0.339940   
min              0.000005            0.015152        0.000046        0.138538   
25%              0.000013            0.040719        0.000075        0.249333   
50%              0.000023            0.066952        0.000106        0.371938   
75%              0.000042            0.136977        0.000191        0.603894   
max              0.000095            0.329891        0.000498        1.667584   

       daily_block_gas_fee_eth  daily_block_gas_fee_usd  \
count                31.000000                31.000000   
mean                  0.148640               472.365303   
std                   0.106105               340.059898   
min                   0.045063               136.460147   
25%                   0.074957               246.873810   
50%                   0.105533               371.937996   
75%                   0.190946               598.080295   
max                   0.498130              1667.583628   

       avg_block_gas_price_gwei  avg_block_utilization  avg_eth_price_usd  \
count                 31.000000              31.000000          31.000000   
mean                   1.352915               0.510027        3208.819608   
std                    0.718361               0.034689         313.694957   
min                    0.580954               0.450364        2747.756250   
25%                    0.860946               0.490394        2985.383958   
50%                    1.131889               0.504186        3151.876667   
75%                    1.548370               0.521875        3437.025833   
max                    3.769551               0.603422        3875.721250   

       max_eth_price_usd  min_eth_price_usd  total_value_eth  total_value_usd  
count          31.000000          31.000000     3.100000e+01     3.100000e+01  
mean         3286.653226        3124.873871     3.695872e+21     1.171921e+25  
std           327.560728         310.975464     6.453924e+21     2.058534e+25  
min          2789.270000        2681.420000     2.855070e+20     8.702222e+23  
25%          3045.015000        2891.630000     5.713862e+20     1.765644e+24  
50%          3214.990000        3060.440000     8.156861e+20     2.690697e+24  
75%          3558.960000        3322.310000     2.364120e+21     7.559000e+24  
max          3908.240000        3852.860000     2.321097e+22     7.732523e+25

In [7]:
import plotly.express as px
import plotly.graph_objects as go

fig = px.line(df, x = "day",
              y = "avg_block_gas_price_gwei",
              title= "Daily Average Gas Price (gwei)")

overall_avg = df["avg_block_gas_price_gwei"].mean()
overall_med = df["avg_block_gas_price_gwei"].median()

fig.add_trace(go.Scatter(
    x = df["day"],
    y = [overall_avg] * len(df),  # this creates a list e.g. [50,50,50,50,50]
    mode= "lines",
    name = "Overall Average",
    line = dict(dash = "dash", width = 2, color = "red")  # Plotly uses a dictionary because one trace can have many line styling options
    
))

fig.add_trace(go.Scatter(
    x=df["day"],
    y=[overall_med] * len(df),
    mode="lines",
    name="Overall Median",
    line=dict(dash="dot", width=2, color = "green")
))

df["rolling_7"] = df["avg_block_gas_price_gwei"].rolling(7).mean()

fig.add_trace(go.Scatter(
    x=df["day"],
    y=df["rolling_7"],
    mode="lines",
    name="7-Day Rolling Average",
    line=dict(width=2)
))

fig.show()

## Observation: Daily Average Block Gas Price (gwei)
* The gas price shows high volatility at the beginning of the month, including a noticeable spike above 3.5 gwei around November 3–5, well above both the overall average (~1.35 gwei) and median (~1.13 gwei).
This suggests a short period of increased network activity or congestion at the beginning of the month.

* After this spike, the gas price stabilizes and remains close to or below the 7-day rolling average, indicating a shift into a lower-congestion regime for the rest of the month.

* The rolling average (purple line) clearly trends downward during the second week of November, moving from ~2 gwei toward ~1.2 gwei. This suggests decreasing pressure on the network as the month progresses.

* Toward the end of November, gas prices dip below both mean and median, indicating consistently low network congestion.

## Key Insight

* The month starts with several congestion events, but the network quickly transitions into a much more stable and low-fee period. The overall shape suggests:
    - Early short-lived spikes → heavy short-term demand
    - Mid- and late-month → predominantly light network usage
* This aligns well with typical Ethereum behavior: brief episodic spikes driven by specific events, surrounded by long periods of moderate-to-low gas fees.
* The month overall shows relatively low gas prices, with most days falling below the monthly mean and median.

In [2]:
from plotly.subplots import make_subplots
import plotly.graph_objects as go

fig = make_subplots(
    rows=2, cols=2,
    subplot_titles=(
        "Daily Gas Price (gwei)",
        "Daily Block Utilization",
        "Median Gas Fee (USD)",
        "ETH Price (USD)"
    )
)

# 1 — Gas Price
fig.add_trace(
    go.Scatter(x=df.day, y=df.avg_block_gas_price_gwei, mode="lines", name="Gas Price"),
    row=1, col=1
)

# 2 — Utilization
fig.add_trace(
    go.Scatter(x=df.day, y=df.avg_block_utilization, mode="lines", name="Utilization"),
    row=1, col=2
)

# 3 — Fee USD
fig.add_trace(
    go.Scatter(x=df.day, y=df.median_gas_fee_usd, mode="lines", name="Median Fee USD"),
    row=2, col=1
)

# 4 — ETH Price
fig.add_trace(
    go.Scatter(x=df.day, y=df.avg_eth_price_usd, mode="lines", name="ETH Price"),
    row=2, col=2
)

fig.update_layout(height=800, width=1200, title_text="Daily Network Metrics Overview")
fig.show()


In [2]:
df[["avg_block_utilization", "avg_block_gas_price_gwei"]].corr(method="pearson")


avg_block_utilization  avg_block_gas_price_gwei
avg_block_utilization                  1.000000                 -0.234623
avg_block_gas_price_gwei              -0.234623                  1.000000

In [3]:
df[["avg_block_utilization", "avg_block_gas_price_gwei"]].corr(method="spearman")


avg_block_utilization  avg_block_gas_price_gwei
avg_block_utilization                  1.000000                 -0.190721
avg_block_gas_price_gwei              -0.190721                  1.000000

## Observation: Block Utilization vs Daily Gas Price (gwei)

* Block utilization throughout November remained within a relatively narrow range, fluctuating between ~48% and ~58%. This indicates that the network operated close to Ethereum’s target utilization under EIP-1559, where blocks are expected to be around 50% full.

* Daily gas prices showed low-to-moderate volatility, ranging between 0.6 gwei and 3.5 gwei, with a brief spike in the first week followed by a stabilization into consistently low values for the remainder of the month.

* When comparing both trends, there is no visually strong alignment between increases in utilization and increases in gas price. Spikes in gas price do not coincide with spikes in utilization, and vice versa.

* Statistical analysis supports this observation:

    * Pearson correlation: −0.23

    * Spearman correlation: −0.19
    Both indicate a weak and statistically insignificant relationship between block utilization and gas price.

 => These results suggest that variations in gas price during November were not driven by block congestion, but instead by short-term mempool activity or other external factors.


 ## Key Insight

* Although Ethereum gas prices can rise sharply during periods of sustained congestion, no such congestion occurred in November. Block utilization stayed close to the 50% target, meaning the baseFee adjustment mechanism under EIP-1559 applied only minimal upward pressure.

* As a result, daily gas prices remained low despite small fluctuations in utilization. The weak correlation confirms that block fullness did not play a meaningful role in gas price movements during this period.

## Observation: Median Gas Fee (USD) vs ETH Price (USD)

* The USD-denominated gas fee shows day-to-day fluctuations that closely resemble the pattern of the daily gas price (gwei).

* Although ETH price declined throughout the month, its movement was gradual and smooth. In contrast, gas price exhibited more noticeable short-term variability.

* As a result, USD gas fees are visually more aligned with gas price than with ETH price, indicating that short-term user costs were driven primarily by changes in gas demand rather than market volatility.

* ETH price did contribute to the overall downward trend in USD gas fees, but it did not explain the day-to-day variations as effectively as the gas price (gwei) did.


## Key Insight

USD gas fees during November were influenced more strongly by fluctuations in gas price (gwei) than by ETH price movements. ETH’s downward trend lowered overall USD fees, but short-term cost spikes and dips were driven almost entirely by gas price variability, not by market conditions.

In [7]:
import plotly.graph_objects as go

fig = go.Figure()

# --- Bar Chart for Transaction Count ---
fig.add_trace(
    go.Bar(
        x=df["day"],
        y=df["daily_block_tx_count"],
        name="Transaction Count",
        marker_color="rgba(55, 83, 109, 0.5)",
        yaxis="y2"   # <-- Assign to secondary axis
    )
)

# --- Line Chart for Gas Price ---
fig.add_trace(
    go.Scatter(
        x=df["day"],
        y=df["avg_block_gas_price_gwei"],
        mode="lines+markers",
        name="Average Gas Price (gwei)",
        line=dict(color="firebrick", width=3),
        yaxis="y1"   # <-- Primary axis
    )
)

# --- Layout / Axes Setup ---
fig.update_layout(
    title="Daily Gas Price (gwei) vs Transaction Count",
    xaxis=dict(title="Day"),
    
    yaxis=dict(
        title="Gas Price (gwei)",
        side="left"
    ),
    
    yaxis2=dict(
        title="Transaction Count",
        overlaying="y",
        side="right"
    ),
    
    bargap=0.3,
    height=500
)

fig.show()


In [8]:
df[["daily_block_tx_count", "avg_block_gas_price_gwei"]].corr(method="pearson")


daily_block_tx_count  avg_block_gas_price_gwei
daily_block_tx_count                  1.000000                 -0.201835
avg_block_gas_price_gwei             -0.201835                  1.000000

## Observation: Daily Gas Price vs Transaction Count

* Daily transaction count in the sampled blocks shows high variability with no clear trend or pattern. Counts fluctuate irregularly between ~8,000 and ~15,000 transactions per day, with no sustained increase or decrease over the period (consider that the dataset is based on a small daily sample of blocks and transactions).

* Gas prices also fluctuate independently, with short-lived price spikes occurring on days that do not correspond to unusually high transaction activity. Likewise, several days with relatively high transaction counts exhibit low gas prices.

* A correlation test confirms this lack of relationship:

    * Pearson correlation: −0.20
    (weak, negative, statistically insignificant)

In [ ]:
import plotly.graph_objects as go

fig = go.Figure()

# --- Line for Gas Price ---
fig.add_trace(
    go.Scatter(
        x=df["day"],
        y=df["avg_block_gas_price_gwei"],
        mode="lines+markers",
        name="Avg Gas Price (gwei)",
        line=dict(color="firebrick", width=3),
        yaxis="y1"
    )
)

# --- Line for Total Value Transferred ---
fig.add_trace(
    go.Scatter(
        x=df["day"],
        y=df["total_value_usd"],
        mode="lines",
        name="Total Value Transferred (USD)",
        line=dict(color="steelblue", width=2),
        yaxis="y2"
    )
)

# --- Layout ---
fig.update_layout(
    title="Daily Gas Price vs Total Value Transferred (USD)",
    xaxis=dict(title="Day"),
    
    yaxis=dict(
        title="Gas Price (gwei)",
        side="left"
    ),
    
    yaxis2=dict(
        title="Total Value (USD)",
        overlaying="y",
        side="right"
    ),
    
    height=500
)

fig.show()


In [10]:
df["day"] = pd.to_datetime(df["day"])
df["weekday"] = df["day"].dt.weekday

# weekday: 0–4, weekend: 5–6
df["is_weekend"] = df["weekday"].apply(lambda x: "Weekend" if x >= 5 else "Weekday")

week_summary = df.groupby("is_weekend")["avg_block_gas_price_gwei"].mean().reset_index()

import plotly.express as px

fig = px.bar(
    week_summary,
    x="is_weekend",
    y="avg_block_gas_price_gwei",
    title="Average Gas Price: Weekdays vs Weekends",
    labels={
        "is_weekend": "",
        "avg_block_gas_price_gwei": "Average Gas Price (gwei)"
    },
    color="is_weekend",
)

fig.update_layout(showlegend=False)
fig.show()


## Observation: Weekday vs Weekend Gas Price

* The data shows a small but noticeable difference between weekday and weekend gas costs.
Average gas prices on weekdays are around 1.5 gwei, whereas weekends average closer to 0.9 gwei.

* This indicates that the network tends to be cheaper to use on weekends. This pattern aligns with typical Ethereum usage behavior.

## Key Insight

Gas prices are meaningfully lower on weekends, reflecting reduced network activity during non-working days.
This observation highlights a behavioral trend consistent with Ethereum usage patterns and reinforces the absence of congestion during the month.

